smart-contracts/ethereum/contracts/MarketPlace.sol

```diff
@@ -1,43 +1,105 @@
-pragma solidity ^0.4.15;
-
-/**
- * The MarketPlace contract:
- * - stores all the information about hyperbridge apps
- */
-contract Marketplace {
-  modifier onlyOperator() {
-    require(operators[msg.sender]);
+pragma solidity >=0.4.15;
+
+contract MarketPlace {
+
+  modifier onlyDelegates() {
+    require(delegates[msg.sender]);
     _;
   }
 
-  struct appVersion {
-    uint version;
+  struct AppVersion {
+    bytes32 version;
     string files;
-    string category;
-    string status;
+    mapping (address => int8) votes;
+    address[] voters; // we might exclude old voters after elections
     string checksum;
-    uint required_permissions;
+    uint permissions;
   }
+  
 
   struct App {
     uint id;
     address owner;
     string name;
-    appVersion[] versions;
+    string category;
+    mapping  (bytes32 => AppVersion)  versions;
+    bytes32[] versionList;
   }
   
   mapping (uint => App) apps;
-  mapping (address => bool) operators;
+  uint[] app_ids;
+  mapping (address => bool) delegates;
+
+  address republicContract;
+  
+  event AppSubmitted(uint id);
+  event VersionSubmitted(uint app_id, bytes32 version);
+  event AppVoting(uint app_id, bytes32 version, int8 vote);
+  
+  function MarketPlace() public {
+    delegates[msg.sender] = true; // only one delegate for now until republic contract is fully functional
```

Comment:
In my latest republic code I prefer to explicitly check if the sender is a delegate or the owner (creator) - I think making the creator a delegate confuses the delegate system.

Reply:
right! we can do that as well